dipy/data/fetcher.py

```diff
@@ -1786,3 +1783,105 @@ def fetch_hcp(subjects,
                            "PipelineDescription": {'Name': 'hcp_pipeline'}})
 
     return data_files, pjoin(my_path, study)
+
+
+def fetch_hbn(subjects, path=None):
+    """
+    Fetches preprocessed data from the Healthy Brain Network POD2 study [1, 2]_.
+
+    Parameters
+    ----------
+    subjects : list
+        Identifiers of the subjects to download.
+        For example: ["NDARAA112DMH", "NDARAA117NEJ"].
+    path : string, optional
+        Path to save files into. Default: '~/AFQ_data'
+
+    Returns
+    -------
+    dict with remote and local names of these files,
+    path to BIDS derivative dataset
+
+    Notes
+    -----
+
+    .. [1] Alexander LM, Escalera J, Ai L, et al. An open resource for
+        transdiagnostic research in pediatric mental health and learning
+        disorders. Sci Data. 2017;4:170181.
+
+    .. [2] Richie-Halford A, Cieslak M, Ai L, et al. An analysis-ready and
+        quality controlled resource for pediatric brain white-matter research.
+        Scientific Data. 2022;9(1):1-27.
+
+    """
+
+    if has_boto3:
+        from botocore import UNSIGNED
+        from botocore.client import Config
+    else:
+        TripWire("The `fetch_hbn` function requires the boto3",
+                 "library, but that is not installed.")
+
+    # Anonymous access:
+    client = boto3.client('s3', config=Config(signature_version=UNSIGNED))
+
+    if path is None:
+        if not op.exists(dipy_home):
+            os.mkdir(dipy_home)
+        my_path = dipy_home
+    else:
+        my_path = path
+
+    base_dir = op.join(my_path, "HBN", 'derivatives', 'qsiprep')
+
+    if not os.path.exists(base_dir):
+        os.makedirs(base_dir, exist_ok=True)
+
+    data_files = {}
+
+    # If user provided incorrect input, these are typical failures that
+    # are easy to recover from:
+    if isinstance(subjects, int) or isinstance(subjects, str):
+        subjects = [subjects]
+
+    for subject in subjects:
+        initial_query = client.list_objects(
+            Bucket="fcp-indi",
+            Prefix=f"data/Projects/HBN/BIDS_curated/sub-{subject}/")
+        ses = initial_query['Contents'][0]["Key"].split('/')[5]
+        query = client.list_objects(
+            Bucket="fcp-indi",
+            Prefix=f"data/Projects/HBN/BIDS_curated/derivatives/qsiprep/sub-{subject}/{ses}/")  # noqa
+        file_list = [kk["Key"] for kk in query["Contents"]]
```

Comment:
Not sure what you are trying to do here.

In my case, it fails:

```python
In [3]: te = fetch_hbn(["NDARAA112DMH", "NDARAA117NEJ"])
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
Cell In[3], line 1
----> 1 te = fetch_hbn(["NDARAA112DMH", "NDARAA117NEJ"])

File ~/devel/dipy/dipy/data/fetcher.py:1855, in fetch_hbn(subjects, path)
   1851 ses = initial_query['Contents'][0]["Key"].split('/')[5]
   1852 query = client.list_objects(
   1853     Bucket="fcp-indi",
   1854     Prefix=f"data/Projects/HBN/BIDS_curated/derivatives/qsiprep/sub-{subject}/{ses}/")  # noqa
-> 1855 file_list = [kk["Key"] for kk in query["Contents"]]
   1856 sub_dir = op.join(base_dir, f'sub-{subject}')
   1857 ses_dir = op.join(sub_dir, ses)

KeyError: 'Contents'
```

Reply:
Additional information:

here what contains `query`:

```
{'ResponseMetadata': {'RequestId': 'N3CXBA1XSTRTR6HE', 'HostId': 'ju6bxhE4xxTXDU0OO7BfObTIefXF97G58A3EstSWmY8Q49OLB8BIxwiNMGWrlrZtAESHlhfxQc0=', 'HTTPStatusCode': 200, 'HTTPHeaders': {'x-amz-id-2': 'ju6bxhE4xxTXDU0OO7BfObTIefXF97G58A3EstSWmY8Q49OLB8BIxwiNMGWrlrZtAESHlhfxQc0=', 'x-amz-request-id': 'N3CXBA1XSTRTR6HE', 'date': 'Thu, 12 Jan 2023 17:13:44 GMT', 'x-amz-bucket-region': 'us-east-1', 'content-type': 'application/xml', 'transfer-encoding': 'chunked', 'server': 'AmazonS3'}, 'RetryAttempts': 0}, 'IsTruncated': False, 'Marker': '', 'Name': 'fcp-indi', 'Prefix': 'data/Projects/HBN/BIDS_curated/derivatives/qsiprep/sub-NDARAA112DMH/ses-HBNsiteRU/', 'MaxKeys': 1000, 'EncodingType': 'url'}
```

Which is strange because `initial_query` work as expected. So either:
- there is an issue in the path 
- there is a limit, cannot have 2 requests in a row so fast.
- or something else.

Reply:
I think this is because the subject ID doesn't exist in the qsiprep derivatives. I believe this is a subject that doesn't even have raw dwi data! https://fcp-indi.s3.amazonaws.com/index.html#data/Projects/HBN/BIDS_curated/sub-NDARAA112DMH/ses-HBNsiteRU/. 

I'll fix the documentation example to point to a subject that has processed data. We might want to think about how we provide information about the full list of subjects that users can point to (it's a long list!). We also have QC data for all these subjects, which users might want to refer to, so we might also want to add a fetcher to that. It's in a csv file in the same s3 bucket.

Reply:
Addressed in f974774f4